extensions/functions/src/main/java/io/crate/window/OffsetValueFunctions.java

```diff
@@ -89,30 +90,119 @@ public Signature boundSignature() {
-            if (offsetValue == null) {
-                return null;
+                          boolean ignoreNulls,
+                          Input[] args) {
+        calculateOffsetIfNull(args);
+        if (offset == null) {
+            return null;
+        }
+        try {
+            if (ignoreNulls == false) {
+                return getValueFromTargetIndex(offsetDirection.getTargetIndex(idxInPartition, offset),
+                                               currentFrame,
+                                               expressions,
+                                               args);
+            }
+            if (offsetDirection == OffsetDirection.FORWARD) {
+                return executeLeadIgnoreNulls(idxInPartition, currentFrame, expressions, args);
             } else {
-                offset = ((Number) offsetValue).intValue();
+                return executeLagIgnoreNulls(idxInPartition, currentFrame, expressions, args);
+            }
+        } catch (IndexOutOfBoundsException e) {
+            return getDefaultOrNull(args);
+        }
+    }
+
+    private Object executeLeadIgnoreNulls(int idxInPartition,
+                                          WindowFrameState currentFrame,
+                                          List<? extends CollectExpression<Row, ?>> expressions,
+                                          Input[] args) {
+        return null;
+    }
+
+    private Object executeLagIgnoreNulls(int idxInPartition,
+                                         WindowFrameState currentFrame,
+                                         List<? extends CollectExpression<Row, ?>> expressions,
+                                         Input[] args) {
+        // cache and idxInPartition forms a window that holds 'offset' number of non-null elements and any number of nulls.
+        // However, if idxInPartition has not moved deeper into the frame, cache will remain null
+
+        // idxInPartition is exclusive so when counting 'offset' number of non-nulls, previous value is used (idxInPartition - 1)
+        var prevValue = getValueFromTargetIndex(idxInPartition - 1, currentFrame, expressions, args);
+        if (prevValue != null) {
+            // idxInPartition is incremented at every iteration and if the prevValue is not null,
+            // cache need to find the idx of the next not null element
+            moveCacheToNextNotNull(idxInPartition, currentFrame, expressions, args);
+        }
+        return getValueFromTargetIndex(cache, currentFrame, expressions, args);
+    }
+
+    private void moveCacheToNextNotNull(int idxInPartition,
+                                        WindowFrameState currentFrame,
+                                        List<? extends CollectExpression<Row, ?>> expressions,
+                                        Input[] args) {
+        if (cache != null) {
+            // from cached index, search in forward direction for a non-null element
+            for (int i = 1; true; i++) {
```

Comment:
Could there be an upper bound? It probably shouldn't move past the current index?

Reply:
it was my assumption that if the cache is not null, the cache and idxInPartition form an offset window such that invoking `moveCacheToNextNotNull` will not exceed idxInPartition but rather adjust to keep my assumption to true.

Also, this method will be used for `lead` as well and it may move past the upperbound of the currentFrame but I am depending on throw mechanism to handle it.